### PR TITLE
planner: replace simple LRU with plan cache LRU

### PR DESCRIPTION
**Pull request by fzzf678**
### What problem does this PR solve?

Replace the simple LRU in general  and prepared plan cache
Issue Number: ref #36598 

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Cache size compare

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

To verify whether the new LRU works, we execute one statement `110000+` times with different parameters and finally check the total number of plan in cache.
The session.tidb_prepared_plan_cache_size is set to `10`
********-----------------------------------------------------

飞书20220901-154226
In the previous LRU, it can not control the cached plan accurately. There are `59000+` plan in cache.
********-----------------------------------------------------

飞书20220901-154232
The dedicated LRU for plan cache cached `10` plans .

### BenchmarkPreparedPointGet

> Old      BenchmarkPreparedPointGet-8      193833     6097 ns/op
> New    BenchmarkPreparedPointGet-8      189016      6084 ns/op

There is no performance loss in the new LRU cache.

### CODE REVIEW

File: planner/core/plan_cache_utils.go [hunk 1]
```diff
@@ -43,38 +43,17 @@ var (
 
 func getValidPlanFromCache(sctx sessionctx.Context, isGeneralPlanCache bool, key kvcache.Key, paramTypes []*types.FieldType) (*PlanCacheValue, bool) {
```
Comment (qw4990) [inline planner/core/plan_cache_utils.go:44]: Seems like we can remove `getValidPlanFromCache` and `putPlanIntoCache` now.

File: planner/core/plan_cache_test.go [hunk 2]
```diff
@@ -95,3 +96,13 @@ func TestGeneralPlanCacheParameterizer(t *testing.T) {
 	tk.MustQuery("select * from t where a > 2 and a < 5").Sort().Check(testkit.Rows("3", "4"))
 	tk.MustQuery("select @@last_plan_from_cache").Check(testkit.Rows("1"))
 }
+
+func TestInitLRUWithSystemVar(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("set @@session.tidb_prepared_plan_cache_size = 0") // MinValue: 1
+	sessionVar := tk.Session().GetSessionVars()
```
Comment (Reminiscent) [inline planner/core/plan_cache_test.go:104]: Please add `select @@session.tidb_prepared_plan_cache_size` to check the value

**Review by Reminiscent — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**